core/src/main/java/org/jruby/RubyModule.java

```diff
@@ -4510,6 +4510,35 @@ public final void setConstantVisibility(Ruby runtime, String name, boolean hidde
         storeConstant(name, entry.value, hidden, entry.getFile(), entry.getLine());
     }
 
+    @JRubyMethod(name = "refinements")
+    public IRubyObject refinements(ThreadContext context) {
+        RubyArray<RubyModule> refinementModules = context.runtime.newArray();
+
+        refinements.forEach((key, value) -> {
+            refinementModules.append(value);
+        });
+        return refinementModules;
+    }
+
+    @JRubyMethod(name = "target")
+    public IRubyObject target(ThreadContext context) {
+        return getRefinedClassOrThrow(context, true);
+    }
+    @JRubyMethod(name = "refined_class")
+    public IRubyObject refined_class(ThreadContext context) {
+        return getRefinedClassOrThrow(context, false);
+    }
+
+    private IRubyObject getRefinedClassOrThrow(ThreadContext context, boolean nameIsTarget) {
+        if (!isRefinement()) {
+            String name = getSimpleName();
+            String clsName = getMetaClass().getSimpleName();
+            String methodName = nameIsTarget? "target":"refined_class";
+            String errMsg = "undefined method `"+ methodName +"' for "+ name +":"+ clsName;
```

Comment:
@enebo might want to comment here; do we need to do anything extra here for encoding? Class names etc.

Reply:
@moste00 @headius yeah something like (static RubyStringBuilder.str):

```
String errMsg = str(runtime, "undefined method `" + methodName + "' for ", rubyBaseName(), ":", getMetaClass());
```

`methodName` is going to be a simple ascii name so we can just j.l.String concat but module name and metaclass name can be multibyte encoded and might not print out properly.  rubyBaseName() is like simpleName but will properly generate an encoded RubyString.  passing in metaclass should have it properly print out the type of the metaclass which if it is within singleton or other patterns will properly print out the string (see RubyStringBuilder.types to see the other processing it will do to types it encounters).'

It is also possible you do not want rubyBaseName and getmetaclass but could use just `this`.   not sure.